File: app/src/main/java/com/h5mota/MainActivity.java
package com.h5mota;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.text.TextUtils;
import android.util.Log;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.EditText;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import org.json.JSONObject;

import java.io.File;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.List;

import fi.iki.elonen.SimpleWebServer;
import me.weyye.hipermission.HiPermission;
import me.weyye.hipermission.PermissionCallback;
import me.weyye.hipermission.PermissionItem;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class MainActivity extends AppCompatActivity {
  public static final String DOMAIN = "https://h5mota.com";
  public static final String LOCAL = "http://127.0.0.1:1055/";

  public File directory;

  SimpleWebServer simpleWebServer;
  SharedPreferences preferences;
  double exittime = 0;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    preferences = getSharedPreferences("config", Context.MODE_PRIVATE);

    List<PermissionItem> list = new ArrayList<>();
    list.add(
        new PermissionItem(
            Manifest.permission.WRITE_EXTERNAL_STORAGE, "存储权限", R.drawable.permission_ic_storage));
    list.add(
        new PermissionItem(
            Manifest.permission.READ_PHONE_STATE, "读取手机状态", R.drawable.permission_ic_phone));

    HiPermission.create(this)
        .title("权限申请")
        .permissions(list)
        .msg("你需要如下权限来使用本软件")
        .checkMutiPermission(
            new PermissionCallback() {
              @Override
              public void onClose() {
                Log.i("Main", "onClose");
              }

              @Override
              public void onFinish() {
                Log.i("Main", "onFinish");
                initSDCard();
              }

              @Override
              public void onDeny(String permission, int position) {
                Log.i("Main", "onDeny");
              }

              @Override
              public void onGuarantee(String permission, int position) {
                Log.i("Main", "onGuarantee");
              }
            });

    findViewById(R.id.online)
        .setOnClickListener(
            view -> loadUrl(DOMAIN, "HTML5魔塔列表"));

    findViewById(R.id.bbs)
        .setOnClickListener(
            view -> loadUrl(DOMAIN + "/bbs", "HTML5bbs"));

    if (!findViewById(R.id.offline).hasOnClickListeners()) {
      findViewById(R.id.offline)
          .setOnClickListener(
              view -> new AlertDialog.Builder(MainActivity.this)
                  .setTitle("错误")
                  .setMessage("你没有SD卡的权限！")
                  .setCancelable(true)
                  .setPositiveButton("确定", null)
                  .create()
                  .show());
    }

    new Thread(
        () -> {
          try {
            OkHttpClient okHttpClient =
                new OkHttpClient()
                    .newBuilder()
                    .followRedirects(true)
                    .followSslRedirects(true)
                    .build();
            try (Response response =
                     okHttpClient
                         .newCall(
                             new Request.Builder()
                                 .url(DOMAIN + "/games/_client/")
                                 .build())
                         .execute()) {
              String s = response.body().string();
              JSONObject jsonObject = new JSONObject(s);
              final JSONObject android = jsonObject.getJSONObject("android");
              String version = android.getString("version");
              if (!version.equals(BuildConfig.VERSION_NAME)) {
                runOnUiThread(
                    () -> {
                      try {
                        new AlertDialog.Builder(MainActivity.this)
                            .setTitle("存在版本更新！")
                            .setMessage(android.getString("text"))
                            .setCancelable(true)
                            .setPositiveButton(
                                "下载",
                                (dialogInterface, i) -> {
                                  try {
                                    loadUrl(android.getString("url"), "版本更新");
                                  } catch (Exception e) {
                                    e.printStackTrace();
                                  }
                                })
                            .setNegativeButton("取消", null)
                            .create()
                            .show();
                      } catch (Exception e) {
                        e.printStackTrace();
                      }
                    });
              }
            }
          } catch (Exception e) {
            e.printStackTrace();
          }
        })
        .start();
  }

  private void initSDCard() {

    // check permission
    if (!HiPermission.checkPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)) return;

    directory = getExternalFilesDir("towers");

    File defaultTower = new File(directory, "24层魔塔");
    if (!defaultTower.exists()) {
      defaultTower.mkdir();
      new Thread(
          () -> Utils.copyFilesFassets(
              MainActivity.this, "24层魔塔", defaultTower.getAbsolutePath()))
          .start();
    }

    testMigration();

    try {
      if (simpleWebServer != null) {
        simpleWebServer.stop();
      }
      simpleWebServer = new MyWebServer("127.0.0.1", 1055, directory, true);
      simpleWebServer.start();
    } catch (Exception e) {
      e.printStackTrace();
      simpleWebServer = null;
    }

    findViewById(R.id.offline)
        .setOnClickListener(
            view -> {
              AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.this);
              builder.setCancelable(true);
              builder.setTitle("已下载的游戏列表");

              final List<String> names = new ArrayList<>();
              names.add("（打开目录...）");
              for (File file : directory.listFiles()) {
                if (new File(file, "index.html").exists()
                    && new File(file, "main.js").exists()
                    && new File(file, "libs").exists()) {
                  names.add(file.getName());
                }
              }

              builder.setItems(
                  names.toArray(new String[0]),
                  (dialogInterface, i) -> {
                    try {
                      if (i == 0) {
                        openDirectory(getExternalFilesDir("towers"));
                      } else {
                        String name = names.get(i);
                        loadUrl(LOCAL + URLEncoder.encode(name, "utf-8"), name);
                      }
                    } catch (Exception e) {
                      e.printStackTrace();
                    }
                  });

              builder.create().show();
            });
  }

  protected void onDestroy() {
    if (simpleWebServer != null) {
      simpleWebServer.stop();
    }
    super.onDestroy();
  }

  public void wantToExit() {
    if (System.currentTimeMillis() - exittime > 2000) {
      Toast.makeText(this, "再按一遍退出程序", Toast.LENGTH_SHORT).show();
      exittime = System.currentTimeMillis();
    } else {
      exittime = 0;
      setResult(RESULT_CANCELED);
      finish();
    }
  }

  @Override
  public boolean onKeyDown(int keyCode, KeyEvent event) {
    if (keyCode == KeyEvent.KEYCODE_BACK && event.getAction() == KeyEvent.ACTION_DOWN) {
      wantToExit();
      return true;
    }
    return super.onKeyDown(keyCode, event);
  }

  public boolean onCreateOptionsMenu(Menu menu) {
    return super.onCreateOptionsMenu(menu);
  }

  public boolean onPrepareOptionsMenu(Menu menu) {
    menu.clear();
    menu.add(Menu.NONE, 0, 0, "")
        .setIcon(android.R.drawable.ic_menu_set_as)
        .setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS);
    menu.add(Menu.NONE, 1, 1, "")
        .setIcon(android.R.drawable.ic_menu_manage)
        .setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS);
    menu.add(Menu.NONE, 2, 2, "")
        .setIcon(android.R.drawable.ic_menu_close_clear_cancel)
        .setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
    return true;
  }

  public boolean onOptionsItemSelected(MenuItem item) {
    super.onOptionsItemSelected(item);
    switch (item.getItemId()) {
      case 0:
        inputLink();
        break;
      case 1:
        new AlertDialog.Builder(this)
            .setSingleChoiceItems(
                new String[]{"启用本地化存档", "禁用本地化存档"},
                preferences.getBoolean("local_save", true) ? 0 : 1,
                (dialogInterface, i) -> {
                  preferences.edit().putBoolean("local_save", i == 0).apply();
                  CustomToast.showSuccessToast(this, i == 0 ? "本地化存档已启用！" : "本地化存档已禁用！");
                }
            )
            .setPositiveButton("确定", null)
            .setCancelable(true).show();
        break;
      case 2:
        finish();
        break;
    }
    return true;
  }

  public void loadUrl(String url, String title) {
    try {
      Intent intent = new Intent(MainActivity.this, WebActivity.class);
      intent.putExtra("title", title);
      intent.putExtra("url", url);
      startActivity(intent);
    } catch (Exception e) {
      e.printStackTrace();
      CustomToast.showErrorToast(this, "无法打开网页！");
    }
  }

  private void inputLink() {
    final EditText editText = new EditText(this);
    editText.setHint("请输入地址...");
    new AlertDialog.Builder(this)
        .setTitle("浏览网页")
        .setView(editText)
        .setPositiveButton(
            "确定",
            (dialogInterface, i) -> {
              String url = editText.getEditableText().toString();
              if (!url.startsWith("http://") && !url.startsWith("https://")) {
                url = "https://" + url;
              }
              loadUrl(url, "浏览网页");
            })
        .setNegativeButton("取消", null)
        .setCancelable(true)
        .create()
        .show();
  }

  private void openDirectory(File directory) {
     Intent intent = new Intent(Intent.ACTION_VIEW);
     intent.setDataAndType(Uri.parse(directory.getAbsolutePath()), "*/*");
     startActivity(Intent.createChooser(intent, "推荐使用ES文件浏览器打开目录..."));
  }

  private void testMigration() {
    if (preferences.getBoolean("migration", false)) {
      return;
    }

    File file = new File(Environment.getExternalStorageDirectory() + "/H5mota/");
    if (!file.exists()) return;

    // test migration
    File[] files = file.listFiles();
    if (files == null || files.length == 0) return;
    List<String> notMigrated = new ArrayList<>();
    for (File f : files) {
      if (!f.isDirectory()) continue;
      if (new File(f, "index.html").exists()
          && new File(f, "main.js").exists()
          && new File(f, "libs").exists()) {
        String name = f.getName();
        File target = new File(directory, name);
        if (!target.exists() && !f.renameTo(target)) {
          notMigrated.add(" - " + name);
        }
      }
    }

    new AlertDialog.Builder(this).setTitle("提示")
        .setMessage("根据Android系统要求，从此版本开始，所有下载的离线塔将挪动至" +
            "SD卡的 /Android/data/com.h5mota/files/towers/ 目录下。\n" +
            (notMigrated.isEmpty() ? "系统已帮你移动了所有离线塔到对应目录。" :
                "系统尝试帮你移动离线塔到对应目录，然而下述塔未移动成功，你需要手动进行移动：\n" +
                TextUtils.join("\n", notMigrated))
        ).setCancelable(false)
        .setNeutralButton("确定，不再提示", (d, i) -> preferences.edit().putBoolean("migration", true).apply())
        .setNegativeButton("打开旧目录", (d, i) -> openDirectory(file))
        .setPositiveButton("打开新目录", (d, i) -> openDirectory(directory))
        .create().show();
  }
}

File: app/src/main/java/com/h5mota/MainActivity.kt
package com.h5mota

import android.app.AlertDialog
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.webkit.WebView
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.windowsizeclass.ExperimentalMaterial3WindowSizeClassApi
import androidx.compose.material3.windowsizeclass.calculateWindowSizeClass
import androidx.compose.ui.Modifier
import com.h5mota.core.server.MotaWebServer
import com.h5mota.core.theme.MotaAndroidTheme
import com.h5mota.core.base.Constant.DIRECTORY
import com.h5mota.core.base.Constant.LOCAL_HOST
import com.h5mota.core.base.Constant.LOCAL_PORT
import com.h5mota.ui.MotaApp
import com.hjq.permissions.OnPermissionCallback
import com.hjq.permissions.Permission
import com.hjq.permissions.XXPermissions
import java.io.File


class MainActivity : ComponentActivity() {
    lateinit var jsActivityLauncher: JsActivityLauncher
    val webScreenLifeCycleHook = WebScreenLifeCycleHook()
    var simpleWebServer: MotaWebServer? = null

    @OptIn(ExperimentalMaterial3WindowSizeClassApi::class)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        jsActivityLauncher = JsActivityLauncher(this)

        if (checkStoragePermission()) initStorage()

        setContent {
            MotaAndroidTheme {
                // A surface container using the 'background' color from the theme
                Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colorScheme.background
                ) {
                    MotaApp(windowSizeClass = calculateWindowSizeClass(this))
                }
            }
        }
    }

    override fun onDestroy() {
        simpleWebServer?.stop()
        simpleWebServer = null
        super.onDestroy()
    }

    override fun onResume() {
        webScreenLifeCycleHook.onResume()
        super.onResume()
    }

    override fun onPause() {
        webScreenLifeCycleHook.onPause()
        super.onPause()
    }

    fun checkStoragePermission(): Boolean {
        return XXPermissions.isGranted(this, Permission.MANAGE_EXTERNAL_STORAGE)
    }

    fun checkNotificationPermission(): Boolean {
        return XXPermissions.isGranted(this, Permission.NOTIFICATION_SERVICE)
    }

    fun requestPermission(callback: () -> Unit) {
        XXPermissions
            .with(this)
            .permission(Permission.NOTIFICATION_SERVICE)
            .permission(Permission.MANAGE_EXTERNAL_STORAGE)
            .request(object : OnPermissionCallback {
                override fun onGranted(
                    permissions: MutableList<String>,
                    allGranted: Boolean
                ) {
                    callback()
                    if (permissions.contains(Permission.MANAGE_EXTERNAL_STORAGE)) initStorage()
                }

                override fun onDenied(
                    permissions: MutableList<String>,
                    doNotAskAgain: Boolean
                ) {
                    callback()
                    noStorageAlert()
                }
            })
    }

    fun noStorageAlert() {
        Toast.makeText(
            this, "无存储卡读写权限，离线游戏暂时无法使用，请在系统设置中启用本软件的存储卡读写权限",
            Toast.LENGTH_LONG
        ).show()
    }

    fun initStorage() {
        if (!checkStoragePermission() || simpleWebServer != null) {
            return
        }

        if (!DIRECTORY.exists()) {
            DIRECTORY.mkdir()
            File(DIRECTORY, ".nomedia").createNewFile()
        } else if (!File(DIRECTORY, ".nomedia").exists()) {
            AlertDialog
                .Builder(this)
                .setTitle("提示")
                .setMessage("这个版本开始，离线塔目录下的图片不会被索引到图库中。\n\n" +
                        "对于已经被索引的图片，可以采用如下方法进行处理：\n" +
                        "1. 使用文件管理器打开手机存储，找到 H5mota 目录\n" +
                        "2. 将其重命名，例如 H5mota1\n" +
                        "3. 打开图库，确认已经被索引的图片消失\n" +
                        "4. 将目录重命名回 H5mota 即可\n" +
                        "如有问题可在论坛或群里进行反馈。")
                .setCancelable(true)
                .setPositiveButton("确定", null)
                .show()
            File(DIRECTORY, ".nomedia").createNewFile()
        }

        try {
            simpleWebServer = MotaWebServer(LOCAL_HOST, LOCAL_PORT, DIRECTORY, true)
            simpleWebServer!!.start()
        } catch (e: Exception) {
            e.printStackTrace()
            simpleWebServer = null
        }

        Thread {
            maybeMigrateLegacyTowers()
        } .start()
    }

    fun maybeMigrateLegacyTowers() {
        val legacyTowerDir = getExternalFilesDir("towers")
        if (legacyTowerDir == null || !legacyTowerDir.exists()) return
        val towers = legacyTowerDir.listFiles()!!
            .filter { tower ->
                tower.isDirectory && File(tower, "index.html").exists()
                        && File(tower, "main.js").exists()
                        && File(tower, "libs").exists()
            }
        towers.forEach { tower ->
            val dest = File(DIRECTORY, tower.name)
            if (dest.exists() || tower.copyRecursively(dest)) {
                tower.deleteRecursively()
            }
        }
    }
}

class WebScreenLifeCycleHook {
    var view: WebView? = null
    private var bgmPaused = false

    val url get() = view?.url

    fun setWebScreen(webView: WebView) {
        view = webView
        bgmPaused = false
    }

    fun onPause() {
        view?.apply {
            pauseMusic()
            onPause()
            pauseTimers()
        }
    }

    fun onResume() {
        view?.apply {
            resumeMusic()
            onResume()
            resumeTimers()
        }
    }

    fun pauseMusic() {
        view?.evaluateJavascript(
            """
                (function() {
                    if (!window.main || !window.core || window.main.mode != 'play') {
                        return false;
                    }
                    if (!window.core.musicStatus.bgmStatus) {
                           return false;
                    }
                    window.core.musicStatus.bgmStatus = false;
                    window.core.pauseBgm();
                    return true;
                })()
            """.trimIndent()
        ) { value -> bgmPaused = value == "true" }
    }

    fun resumeMusic() {
        if (bgmPaused) {
            bgmPaused = false
            view?.evaluateJavascript(
                """
                    (function() {
                        if (!window.main || !window.core || window.main.mode != 'play') {
                            return;
                        }
                        if (window.core.musicStatus.bgmStatus) {
                            return;
                        }
                        window.core.musicStatus.bgmStatus = true;
                        window.core.resumeBgm();
                    })()
                """.trimIndent()
            ) {}
        }
    }
}

class JsActivityLauncher(activity: ComponentActivity) {
    private lateinit var callback: (result: ActivityResult) -> Unit
    private val activityResultLauncher = activity.registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result -> callback(result) }

    fun launch(intent: Intent, callback: (result: ActivityResult) -> Unit) {
        this.callback = callback
        activityResultLauncher.launch(intent)
    }
}